[nnc] Make 64-bit dimensions work

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64077
* #64076

We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

bertmaher committed:
[nnc] Make 64-bit dimensions work

We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Make 64-bit dimensions work"


We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Make 64-bit dimensions work"


We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Make 64-bit dimensions work"


We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30596689).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.h:
```diff
@@ -29,6 +29,18 @@ class Value {
+  template <typename T>
+  Value(Dtype dtype, T v) : dtype_(dtype) {
+#define TYPE_CASE(Type, Name)  \
+  if (dtype == k##Name) {      \
```
I don't remember details of the dtypes, but is there any difference between `dtype == k##Name` and `dtype.scalar_type() == ScalarType::Name`, which is used in lots of others places?

bertmaher replied:
Yeah, thk `k##Name`s are all lanes=1, whereas checking the scalar type could let through vectorized types.  The latter *does* look cleaner (I didn't actually think of it at the time) but it does actually feel more accurate to use `==k##Name` since these are scalar constructors.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -664,7 +667,7 @@ std::vector<ExprHandle> TensorExprKernel::sizesForValue(
   if (v->type()->isSubtypeOf(FloatType::get()) ||
       v->type()->isSubtypeOf(IntType::get())) {
-    return {1};
+    return {int64_t{1}};
```
Is it the same as 1LL?

bertmaher replied:
`1ll` is probably fine although I'm not 100% sure.  There are a bunch of weirdnesses around implicit conversions, and the fact that on windows "long" and "long long" have different widths.  `int64_t` seemed like a good way to sidestep all that :-p

ZolotukhinM commented on torch/csrc/jit/tensorexpr/expr.h:
```diff
@@ -319,7 +319,7 @@ class TORCH_API BufHandle : public ExprHandle {
 class TORCH_API VarHandle : public ExprHandle {
  public:
-  VarHandle() : ExprHandle(nullptr) {}
+  VarHandle() : ExprHandle() {}
```
Ah, I think @navahgar fixed it already in #64117.

ZolotukhinM reviewed (approved):
Looks good!

bertmaher commented on benchmarks/cpp/tensorexpr/bench_ops.py:
```diff
@@ -2,7 +2,7 @@
 import torch.nn.functional as F
 
-torch._C._jit_override_can_fuse_on_cpu(True)
+torch._C._jit_override_can_fuse_on_gpu(True)
```
Oh yeah, no, I meant to revert this.  I was benchmarking gpu ops to make sure that using int64 indexing didn't slow us down (it doesn't, at least on an A100).

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30596689).

bertmaher committed:
Update on "[nnc] Make 64-bit dimensions work"


We were assuming kernel dimensions fit in 32 bits (the old fuser made
this assumption too), but we should be able to support 64.

Differential Revision: [D30596689](https://our.internmc.facebook.com/intern/diff/D30596689/)

[ghstack-poisoned]
